exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,251 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+		return err
+	}
+
+	if err := r.deleteDanglingMachines(ctx, mp, machineList.Items); err != nil {
+		return errors.Wrapf(err, "failed to clean up orphan machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	if err := r.createMachinesIfNotExists(ctx, mp, machineList.Items, infraMachineList.Items); err != nil {
+		return errors.Wrapf(err, "failed to create machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	if err := r.setInfraMachineOwnerRefs(ctx, mp, infraMachineList.Items); err != nil {
+		return errors.Wrapf(err, "failed to create machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	return nil
+}
+
+// deleteDanglingMachines deletes all MachinePool Machines with an infraRef pointing to an infraMachine that no longer exists.
+// This allows us to clean up Machines when the MachinePool is scaled down.
+func (r *MachinePoolReconciler) deleteDanglingMachines(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Deleting orphaned machines", "machinePool", mp.Name, "namespace", mp.Namespace)
+
+	for i := range machines {
+		machine := &machines[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will complain if the infraRef is empty, i.e. no object kind or apiVersion.
+		// We want to delete since an empty infraRef is invalid as well.
+		if (infraRef == corev1.ObjectReference{}) {
+			log.V(2).Info("Machine has an empty infraRef, will delete", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		_, err := external.Get(ctx, r.Client, &infraRef, mp.Namespace)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Error(err, "unexpected error while trying to get infra machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
+			}
+
+			log.V(2).Info("Deleting orphaned machine", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+		} else {
+			log.V(2).Info("Machine is not orphaned, nothing to do", "machine", machine.Name, "namespace", machine.Namespace)
+		}
+	}
+
+	return nil
+}
+
+// createMachinesIfNotExists creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
+func (r *MachinePoolReconciler) createMachinesIfNotExists(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine, infraMachines []unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	infraRefNames := sets.Set[string]{}
+	for _, machine := range machines {
+		infraRef := machine.Spec.InfrastructureRef
+		infraRefNames.Insert(infraRef.Name)
+	}
+
+	for i := range infraMachines {
```

Comment:
I have to index it from the array or I get this linting error otherwise:

```
exp/internal/controllers/machinepool_controller_phases.go:463:32: G601: Implicit memory aliasing in for loop. (gosec)
                machine := getNewMachine(mp, &infraMachine)
```